proposals/0010-resource-element-type-validation.md

```diff
@@ -0,0 +1,221 @@
+It will use `BuildFlattenedTypeList` to retrieve a small vector of the subelement types.
+From this subvector, the first element will be compared to all elements in the vector,
+and any mismatches will return false. Typed buffer RETs will 
+also need to have at most 4 subelements, and the total size in bytes cannot exceed 16,
+which will also be verified by `__builtin_hlsl_is_line_vector_layout_compatible`.
+Finally, there will be an additional check that there are no bools or enums present
+in any component of the type.
+
+* Examples of RET validation results:
+```
+
+struct oneInt {
+    int i;
+};
+
+struct twoInt {
+   int aa;
+   int ab;
+};
+
+struct threeInts {
+  oneInt o;
+  twoInt t;
+};
+
+struct oneFloat {
+    float f;
+};
+struct notComplete;
+struct depthDiff {
+  int i;
+  oneInt o;
+  oneFloat f;
+};
+
+struct notHomogenous{     
+  int i;
+  float f;
+};
+
+struct EightElements {
+  twoInt x[2];
+  twoInt y[2];
+};
+
+struct EightHalves {
+half x[8]; 
+};
+
+struct intVec {
+  int2 i;
+};
+
+struct oneIntWithVec {
+  int i;
+  oneInt i2;
+  int2 i3;
+};
+
+struct weirdStruct {
+  int i;
+  intVec iv;
+};
+
+RWBuffer<double2> r0; // valid - RET fits in 4 32-bit quantities
+RWBuffer<int> r1; // valid
+RWBuffer<float> r2; // valid
+RWBuffer<float4> r3; // valid
+RWBuffer<notComplete> r4; // invalid - the RET isn't complete, the definition is missing. 
+// the type trait that would catch this is the negation of `__is_intangible`
+RWBuffer<oneInt> r5; // valid - all leaf types are valid primitive types, and homogenous
+RWBuffer<oneFloat> r6; // valid
+RWBuffer<twoInt> r7; // valid
+RWBuffer<threeInts> r8; // valid
+RWBuffer<notHomogenous> r9; // invalid, all template type components must have the same type, DXC fails
+StructuredBuffer<notHomogenous> r9Structured; // valid
+RWBuffer<depthDiff> r10; // invalid, all template type components must have the same type, DXC fails
+RWBuffer<EightElements> r11; // invalid, > 4 elements and > 16 bytes, DXC fails 
+// This would be caught by __builtin_hlsl_is_line_vector_layout_compatible
+StructuredBuffer<EightElements> r9Structured; // valid
+RWBuffer<EightHalves> r12; // invalid, > 4 elements, DXC fails
+// This would be caught by __builtin_hlsl_is_line_vector_layout_compatible
+StructuredBuffer<EightHalves> r12Structured; // valid
+RWBuffer<oneIntWithVec> r13; // valid
+RWBuffer<weirdStruct> r14; // valid
+RWBuffer<RWBuffer<int> > r15; // invalid - the RET has a handle with unknown size, thus it is an intangible RET.
+// the type trait that would catch this is the negation of `__is_intangible`
+```
+
+Below is a sample C++ implementation of the `RWBuffer` resource type, which is a typed buffer variant.
+This code would exist within an hlsl header, but concepts are not implemented in HLSL. However, the AST node
+associated with RWBuffers is constructed as if this code was read and parsed by the compiler.
+```
+#include <type_traits>
+
+namespace hlsl {
+
+
+template<typename T>

```

Comment:
Looks like something got edited away here.

I expect you'll want to think of this as something like:
```c++
template<typename RET>
concept TypedResourceElementType = 
    __builtin_hlsl_is_line_vector_layout_compatible<RET>() &&
    !std::is_enum_v<RET> && !std::is_same_v<RET, bool>;


template<typename RET>
    requires TypedResourceElementType<RET>
struct Buffer 
{
};
```

(https://godbolt.org/z/MdfYhYqh9)